### PR TITLE
[Federation] Unjoin only the joined clusters while bringing down the federation control plane.

### PR DESCRIPTION
**Pull request by madhusudancs**
A few other minor improvements.

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by madhusudancs** [comment]
@k8s-bot federation gce e2e test this

**Comment by madhusudancs** [comment]
@k8s-bot federation gce e2e test this

**Comment by shashidharatd** [comment]
@madhusudancs, pre-submit pr tests seems to be failing.
```
I0309 17:30:55.995] Extracting kubernetes-test.tar.gz into /workspace/kubernetes
W0309 17:31:02.791] 2017/03/09 17:31:02 util.go:110: Step './get-kube.sh' finished in 15.948444991s
W0309 17:31:02.791] 2017/03/09 17:31:02 util.go:108: Running: ./federation/cluster/federation-down.sh
W0309 17:31:02.828] Project: k8s-jkns-pr-bldr-e2e-gce-fdrtn
W0309 17:31:02.829] Zone: us-central1-f
I0309 17:31:02.993] +++ [0309 17:31:02] Unjoining cluster "federation-e2e-gce-us-central1-b" from federation "e2e-federation"
**W0309 17:31:03.100] error: context "e2e-federation" does not exist**
W0309 17:31:03.104] 2017/03/09 17:31:03 util.go:110: Step './federation/cluster/federation-down.sh' finished in 312.820638ms
W0309 17:31:03.134] 2017/03/09 17:31:03 main.go:153: Saved XML output to /workspace/_artifacts/junit_runner.xml.
W0309 17:31:03.135] 2017/03/09 17:31:03 main.go:187: Something went wrong: error tearing down previous federation control plane: exit status 1
W0309 17:31:03.136] + chmod -R o+r /workspace/_artifacts
I0309 17:31:04.039] Starting pull-kubernetes-federation-e2e-gce-59...
W0309 17:31:04.071] Traceback (most recent call last):
W0309 17:31:04.071]   File "/var/lib/jenkins/workspace/pull-kubernetes-federation-e2e-gce/./test-infra/jenkins/../scenarios/kubernetes_e2e.py", line 409, in <module>
W0309 17:31:04.071]     main(create_parser().parse_args())
W0309 17:31:04.071]   File "/var/lib/jenkins/workspace/pull-kubernetes-federation-e2e-gce/./test-infra/jenkins/../scenarios/kubernetes_e2e.py", line 346, in main
W0309 17:31:04.072]     mode.start(runner_args)
W0309 17:31:04.072]   File "/var/lib/jenkins/workspace/pull-kubernetes-federation-e2e-gce/./test-infra/jenkins/../scenarios/kubernetes_e2e.py", line 239, in start
W0309 17:31:04.072]     check(*cmd)
W0309 17:31:04.072]   File "/var/lib/jenkins/workspace/pull-kubernetes-federation-e2e-gce/./test-infra/jenkins/../scenarios/kubernetes_e2e.py", line 39, in check
W0309 17:31:04.072]     subprocess.check_call(cmd)
W0309 17:31:04.072]   File "/usr/lib/python2.7/subprocess.py", line 540, in check_call
W0309 17:31:04.072]     raise CalledProcessError(retcode, cmd)
```
Should it be fixed in this pr?

**Comment by madhusudancs** [comment]
> Should it be fixed in this pr?

Don't know yet. It depends on where the problem is. If it is in test-infra, then there isn't much we can do here. But let's keep this PR on hold for now. I will use this PR as a guinea pig.

**Comment by madhusudancs** [comment]
@shashidharatd I think this is good to go now. Still won't pass federation presubmit. But there are other changes in test-infra to make that happen. Could review this and if you think it's alright please LGTM?

I am preparing the test-infra side of changes.

**Comment by madhusudancs** [comment]
test-infra PR here - https://github.com/kubernetes/test-infra/pull/2221

**Comment by shashidharatd** [comment]
Jenkins GCE etcd3 e2e

**Comment by madhusudancs** [comment]
@k8s-bot gce etcd3 e2e test this

File: federation/cluster/federation-down.sh [hunk 1]
```diff
@@ -24,9 +24,29 @@ KUBE_ROOT=$(readlink -m $(dirname "${BASH_SOURCE}")/../../)
 # $KUBEDNS_CONFIGMAP_NAME and $KUBEDNS_CONFIGMAP_NAMESPACE.
 source "${KUBE_ROOT}/federation/cluster/common.sh"
 
+# federation_clusters returns a list of all the clusters in
+# federation, if at all the federation control plane exists
+# and there are any clusters registerd.
+function federation_clusters() {
```
Comment (eparis) [inline federation/cluster/federation-down.sh]: Should we really have to hit the API twice? Why not something that only hits the API once?

```bash
if clusters=$("${KUBE_ROOT}/cluster/kubectl.sh" --context="${FEDERATION_KUBE_CONTEXT}" -o jsonpath --template '{.items[*].metadata.name}' get clusters); then
  echo ${clusters}
  return
fi
echo ""
```

Comment (madhusudancs) [inline federation/cluster/federation-down.sh]: @eparis Thanks! Done.

**Comment by madhusudancs** [comment]
@k8s-bot gce etcd3 e2e test this

**Pull request merged**